File: examples/basic_usage.py
#!/usr/bin/env python3
"""
Basic usage examples for Multi-AI MCP Integration
"""

import asyncio
import os
import sys

# Add src to path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'src'))

from multi_ai_integration import MultiAIIntegration, MultiAIConsultationRequest, AIProvider

async def basic_consultation_example():
    """Example: Basic AI consultation with automatic provider selection"""
    print("🤖 Basic AI Consultation Example")
    print("=" * 50)
    
    # Initialize the multi-AI system
    multi_ai = MultiAIIntegration()
    
    # Create a consultation request
    request = MultiAIConsultationRequest(
        query="What are the best practices for API authentication?",
        context="Building a RESTful API for a web application",
        preferred_provider=AIProvider.AUTO,
        compare_responses=False
    )
    
    # Perform consultation
    result = await multi_ai.consult_multi_ai(request)
    
    if result["success"]:
        print("✅ Consultation successful!")
        print(f"📝 Summary:\n{result['summary']}")
    else:
        print(f"❌ Consultation failed: {result['error']}")

async def provider_specific_example():
    """Example: Consulting specific AI providers"""
    print("\n🎯 Provider-Specific Consultation Example")
    print("=" * 50)
    
    multi_ai = MultiAIIntegration()
    
    # Consult Gemini specifically
    print("📡 Consulting Gemini...")
    gemini_request = MultiAIConsultationRequest(
        query="How should I structure a microservices architecture?",
        context="Designing a scalable e-commerce platform",
        preferred_provider=AIProvider.GEMINI
    )
    
    gemini_result = await multi_ai.consult_multi_ai(gemini_request)
    
    if gemini_result["success"]:
        print("✅ Gemini consultation successful!")
        print(f"📄 Response preview: {gemini_result['summary'][:200]}...")
    else:
        print(f"❌ Gemini consultation failed: {gemini_result['error']}")
    
    # Consult Llama2 specifically
    print("\n🦙 Consulting Llama2...")
    llama_request = MultiAIConsultationRequest(
        query="What's a privacy-focused approach to user data handling?",
        context="Implementing GDPR-compliant data processing",
        preferred_provider=AIProvider.LLAMA2
    )
    
    llama_result = await multi_ai.consult_multi_ai(llama_request)
    
    if llama_result["success"]:
        print("✅ Llama2 consultation successful!")
        print(f"📄 Response preview: {llama_result['summary'][:200]}...")
    else:
        print(f"❌ Llama2 consultation failed: {llama_result['error']}")

async def system_status_example():
    """Example: Checking system status and performance"""
    print("\n📊 System Status Example")
    print("=" * 50)
    
    multi_ai = MultiAIIntegration()
    
    # Check provider availability
    availability = await multi_ai.check_provider_availability()
    print("🔌 Provider Availability:")
    for provider, available in availability.items():
        status = "✅ Available" if available else "❌ Unavailable"
        print(f"   • {provider.title()}: {status}")
    
    # Get comprehensive system status
    status = multi_ai.get_system_status()
    print("\n📈 System Statistics:")
    for provider, info in status["providers"].items():
        if info["available"]:
            stats = info.get("stats", {})
            requests = stats.get("requests", 0)
            successes = stats.get("successes", 0)
            avg_time = stats.get("avg_time", 0.0)
            success_rate = (successes / requests * 100) if requests > 0 else 0
            
            print(f"   • {provider.title()}:")
            print(f"     - Requests: {requests}")
            print(f"     - Success Rate: {success_rate:.1f}%")
            print(f"     - Avg Response Time: {avg_time:.2f}s")

async def main():
    """Run all examples"""
    print("🚀 Multi-AI MCP Integration - Usage Examples")
    print("=" * 60)
    
    try:
        await basic_consultation_example()
        await provider_specific_example()
        await system_status_example()
        
        print("\n🎉 All examples completed successfully!")
        print("\nNext steps:")
        print("• Try these patterns in Claude Code")
        print("• Experiment with different query types")
        print("• Monitor system performance over time")
        
    except KeyboardInterrupt:
        print("\n🛑 Examples interrupted")
    except Exception as e:
        print(f"\n❌ Example failed: {e}")

if __name__ == "__main__":
    # Set up environment
    if not os.getenv('GOOGLE_API_KEY'):
        print("⚠️  GOOGLE_API_KEY not set. Some examples may fail.")
        print("   Set it with: export GOOGLE_API_KEY='your-key-here'")
    
    asyncio.run(main())

File: examples/basic-usage.js
#!/usr/bin/env node

/**
 * Basic Usage Examples for Multi-AI Integration
 * Demonstrates core functionality and features
 */

import MultiAI from '../src/index.js';

async function runExamples() {
  console.log('🚀 Multi-AI Integration - Basic Usage Examples\n');

  // Initialize the AI system
  const ai = new MultiAI();
  
  try {
    // Initialize providers
    console.log('1. Initializing AI providers...');
    await ai.initializeProviders();
    
    // Example 1: Basic chat
    console.log('\n2. Basic Chat Example:');
    try {
      const response = await ai.chat('Hello! Tell me a fun fact about programming.');
      console.log(`Response: ${response.response}`);
      console.log(`Provider: ${response.provider}, Model: ${response.model}`);
    } catch (error) {
      console.log('⚠️  Chat failed (no providers available):', error.message);
    }

    // Example 2: Task-specific chat
    console.log('\n3. Task-Specific Chat Example:');
    try {
      const codeResponse = await ai.chat(
        'Write a simple Python function to calculate fibonacci numbers',
        { taskType: 'code' }
      );
      console.log(`Code Response: ${codeResponse.response.substring(0, 200)}...`);
      console.log(`Provider: ${codeResponse.provider}`);
    } catch (error) {
      console.log('⚠️  Code chat failed:', error.message);
    }

    // Example 3: Knowledge base usage
    console.log('\n4. Knowledge Base Example:');
    ai.addKnowledge('project_info', {
      name: 'Multi-AI Integration',
      version: '2.0.0',
      description: 'Advanced multi-provider AI system'
    });
    
    ai.addKnowledge('best_practices', 'Always validate user input and sanitize API keys');
    
    console.log(`Knowledge entries: ${ai.knowledgeBase.size}`);
    
    // Search knowledge
    const searchResults = ai.searchKnowledge('project');
    console.log(`Search results for 'project': ${searchResults.length} entries`);

    // Example 4: Context management
    console.log('\n5. Context Management Example:');
    ai.clearContext();
    ai.updateContext('What is AI?', 'AI stands for Artificial Intelligence...');
    ai.updateContext('Tell me about machine learning', 'Machine learning is a subset of AI...');
    
    const context = ai.getContext();
    console.log(`Context entries: ${context.length}`);
    console.log(`Latest context: ${context[context.length - 1]}`);

    // Example 5: System status
    console.log('\n6. System Status Example:');
    const status = await ai.getSystemStatus();
    console.log(`System version: ${status.version}`);
    console.log(`Healthy providers: ${status.providers.healthy}/${status.providers.total}`);
    console.log(`Knowledge base entries: ${status.resources.knowledgeBase.entries}`);
    console.log(`Memory usage: ${Math.round(status.resources.knowledgeBase.memoryUsage.heapUsed / 1024 / 1024)}MB`);

    // Example 6: Configuration
    console.log('\n7. Configuration Example:');
    console.log(`Prefer local providers: ${ai.config.routing.preferLocal}`);
    console.log(`Max context length: ${ai.config.context.maxLength}`);
    console.log(`Request timeout: ${ai.config.performance?.requestTimeout || 'default'}ms`);

    // Example 7: Error handling
    console.log('\n8. Error Handling Example:');
    try {
      await ai.chat(''); // This should fail
    } catch (error) {
      console.log(`✅ Properly caught error: ${error.message}`);
    }

    try {
      ai.addKnowledge('', 'empty key'); // This should fail
    } catch (error) {
      console.log(`✅ Properly caught error: ${error.message}`);
    }

    // Example 8: Provider statistics (if any requests were made)
    console.log('\n9. Provider Statistics Example:');
    const stats = ai.router.getProviderStats();
    for (const [provider, data] of Object.entries(stats)) {
      if (data.requests > 0) {
        console.log(`${provider}: ${data.requests} requests, ${data.successRate} success rate`);
      }
    }

    console.log('\n✅ All examples completed successfully!');
    console.log('\nNext steps:');
    console.log('- Try the CLI: npm run start');
    console.log('- Process a file: npm run start file ./examples/basic-usage.js');
    console.log('- Check health: npm run health');

  } catch (error) {
    console.error('❌ Example failed:', error.message);
    
    if (error.message.includes('No AI providers')) {
      console.log('\n💡 Tip: Make sure Ollama is installed and running:');
      console.log('   1. Install Ollama: https://ollama.ai/');
      console.log('   2. Pull a model: ollama pull llama3.2');
      console.log('   3. Run this example again');
    }
  }
}

// Run examples if this file is executed directly
if (import.meta.url === `file://${process.argv[1]}`) {
  runExamples().catch(error => {
    console.error('Examples failed:', error);
    process.exit(1);
  });
}

export { runExamples };